### ISSUE TITLE
Covox Sound Master support

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#2219]
# Overview 

Adding support for this card has already been brought up in https://github.com/dosbox-staging/dosbox-staging/issues/1843 

The Covox Sound Master is an internal 8-bit ISA sound card. Its audio capabilities come from an **AY8930/P which is used for music** and an 8-bit mono DAC for digital sound effects and speech. The AY8930/P chip features 3 pulse wave channels and 1 noise channel, each channel is highly configurable, but most developers used its Compatible Mode making it functionally equivalent to the older, but more familiar, AY-3-8910. ([source](http://www.vgmpf.com/Wiki/index.php?title=Sound_Master))

See more details here:

- [Covox Sound Master](http://www.vgmpf.com/Wiki/index.php?title=Sound_Master)
- [AY8930/P](http://www.vgmpf.com/Wiki/index.php?title=AY8930/P)

# Game support

[Source](http://nerdlypleasures.blogspot.com/2012/04/)

### AY8910 or 8930 Sound

- Conan - The Cimmerian
- Galleons of Glory
- Megatraveller 2
- Prince of Persia (v1.0 only)
- Punisher, The
- Sim Earth
- Spirit of Excalibur
- Twilight 2000
- Ultima VI
- Vengeance of Excalibur
- Where in Time if Carmen Sandiego (AY8930 music)
- Where in the World is Carmen Sandiego (320x200 VGA Version)

These games support music and/or sound effects using the AY8930 PSG found on the original Covox Sound Master.  Mostly games only supported the basic AY8910 features, which were also found in the Atari ST computers, many arcade games, the Intellivision and the Mockingboard line of sound cards for the Apple II.  The card is incredibly rare.  Covox marketed a line of sound adapters, including the Speech Thing, the Voice Master, the Sound Master, the Sound Master + and the Sound Master II.  The + and II are NOT compatible with the original Sound Master, as they are essentially Sound Blaster clones.

### Near Exclusive Digital Sound Support

- MegaTraveller 1
- Space 1899
- SimCity
- The Punisher
- Transylvania III

These games support one of the digitial sound reproduction methods of the original Covox Sound Master or the Adlib-compatible Covox Sound Master II and do not support Sound Blaster digitized sound playback.

Note that Covox Lemmings only supports Adlib, Countdown and A-Train does not actually function with a Sound Master (despite what it says on the A-Train box) and Wolfenstein 3D specifically requires a Disney Sound Source and will not work with a Covox Speech Thing/Voice Master/Sound Master of any kind.

# Implementations

The Furnace chiptune trackers has AY8930 support:
https://github.com/tildearrow/furnace

MAME also contains an emulation of the chip, apparently based on measurements:
https://github.com/mamedev/mame/blob/master/src/devices/sound/ay8910.cpp

# More related info

https://www.vogons.org/viewtopic.php?t=18571

**Comment by kcgen** [comment]
https://www.vogons.org/viewtopic.php?t=13804

**Comment by johnnovak** [comment]
> https://www.vogons.org/viewtopic.php?t=13804

Thanks, I've added it to the top post.

**Comment by MasterO2** [comment]
The Covox Emulation in MAME is primarily for emulating the title character's voice from the **Q-Bert** arcade game. 

Q-Bert also has a DOS port, but I dunno if that uses Covox or not.

**Comment by CarlosTex** [comment]
IIRC, DIGPAK (John Ratcliff) library has a driver for the Covox Sound Master DAC (SMSND.COM), so technically every game that uses DIGPAK for digital sound will technically work with the Covox Sound Master, renaming SMSND.COM to SBLASTER.COM could eventually do the trick. HEX editing to select SMSND.COM directly can work too. This increases dramatically the number of games that work with the Covox Sound master DAC.

Also, DIGPAK is compatible with Audio Interface Library 2 (John Miles) starting from AIL2 version 2.07, so the number of games supported increases even more! I made Dune 2 (which uses AIL2) work with the Covox Speech Thing DIGPAK driver (CVXSND.COM), by renaming SBDIG.ADV to SBDIG.ORI (ORI as in original driver) and CVXSND.COM as SBDIG.ADV. Digital sound could be heard from my Covox Speech Thing just fine. This was all tested in real hardware, but should work in emulation too.

**Comment by johnnovak** [comment]
@CarlosTex We're more interested in the other direction: are there any games that have something unique on the Covox? And if so, can a driver hack like that make them run on a SB, making supporting the Covox Sound Master completely unnecessary?

We don't care about emulating the hardware much, just the games that might have something unique on the Covox Sound Master so people can experience that _somehow_.

Probably it's bottom of the barrel stuff and all games that support the CSM support the SB DAC anyway.

**Comment by CarlosTex** [comment]
@johnnovak 

Stuff/games that use Covox Sound Master exclusively for digital sound does not use any middleware that would allow for what i was suggesting. But if one game did use DIGPAK for sound and it shipped with the SMSND.COM exclusively i'm sure it could be made to work with the sound blaster driver or any other driver for any other card that DIGPAK supports. That's the beauty of middleware. I do not know of any.

The same could be said for MIDPAK, which is just a stub that loads the AIL2 music driver accordingly. Legend of Kyrandia (AIL2 game) for instance never supported the Innovation SSI as there was never a AIL2 driver made for the Innovation SSI. But a few years ago a VOGONS member made one so this is how Legend of Kyrandia sounds roughly on a SID chip:

https://www.youtube.com/watch?v=_0uUVfC7e94

I do understand that this is all irrelevant for the DOSBox staging project as you are not interested in emulating more DOS era hardware, so i guess i can't say or help you with anything else. 

I guess that the "policy" around here is that not worth to emulate more than one (except ULTRASOUND) PCM card? Amount of work / reward? The real AdLib Gold card for instance also routes the PCM sound through the effects board for instance, and although no game supports PCM + effects it can be set manually on real hardware.

I would like to help to confirm the exclusive support for the Covox Sound Master games that are claimed, but unfortunately i don't have a card or no clone is available at the moment to confirm that.

**Comment by johnnovak** [comment]
> I guess that the "policy" around here is that not worth to emulate more than one (except ULTRASOUND) PCM card? Amount of work / reward?

The goal is to get games running, and let users experience all unique game features. GUS is important as many games have unique sound that's only achievable with a GUS. Also, most demoscene prods need a GUS (although demoscene is a secondary focus at best).

PCM is pretty much PCM, so if a game supports SB, then any number of other PCM boards that play the same PCM sample the same way with the same sound—well, who cares? We're pragmatic here, the goal is to experience games in all their unique ways. It makes no difference through what emulated interface those PCM samples get pushed through (GUS is an exception, of course, and MOD music played through the GUS sounds a lot better than most crappy CPU-based mixer implementations that the SB drivers use).

> The real AdLib Gold card for instance also routes the PCM sound through the effects board for instance, and although no game supports PCM + effects it can be set manually on real hardware.

We have the `reverb` and `chorus` settings to achieve that in lot more flexible way. Yeah, you can route OPL through the AWE32's chorus/reverb fx on certain AWE boards too—similar thing.

**Comment by CarlosTex** [comment]
> PCM is pretty much PCM, so if a game supports SB, then any number of other PCM boards that play the same PCM sample the same way with the same sound—well, who cares?

Preservationists. I could argue that sometimes PCM played by different sound devices provides a different experience, just like the GUS for the better, but sometimes it might be fun to try it a little worse. For instance, i hacked the Disney Sound System driver for Infogrames games (like Alone in the Dark) so that PCM sound would play on a Covox Speech Thing. PCM is PCM sure but it won't sound exactly the same. With emulation, yeah maybe it's harder to justify but i hope i'm getting my point across. 

> We're pragmatic here

Look, i understand. This is your project afterall, not trying to twist anybody's arm, you have your formula and it works and DOSBox Staging seems to be a great option to play games on a modern system. I left my first comment because i thought my input could eventually be valuable in some way.

Keep up the good work!! If i'm ever able to give some more input on Coox Sound Master exclusive games, i promise i will.

**Comment by johnnovak** [comment]
> Preservationists. I could argue that sometimes PCM played by different sound devices provides a different experience, just like the GUS for the better, but sometimes it might be fun to try it a little worse.

Yeah, I understand your point. When you deal with real hardware, all little details that make no difference in a "digital emulation" start to matter. Like even the label on the driver installation floppy disks, just the feeling of holding it in your hand, etc. Like I'm particular to Matrox graphics adapters in my Win98 boxes, just because that's what I had back in the day. Just makes me feel good 😄 

So yeah, by "PCM is PCM" I meant that in a purely digital emulation (like in most emulators), there is hardly any difference. But I added analog output filter emulation to all audio devices (`sb_filter`, `gus_filter`, etc.), that makes the differences between different audio devices with different output filters more appreciable. Then there's also the different sample rates different PCM devices support. E.g. a really low rate device like the Covox-style devices combined with a specific output filter can yield different sounding results than an SB. And so on.

Case in point, I've managed to emulate the filter and DAC differences between SB16 and earlier SB models quite faithfully (compared to actual recording, it's hard to distinguish the real recording and Staging's output when A/B comparing recordings). It's subtle to most people, but I care about this stuff, yeah, and SB boards are the standard for most DOS games, so kinda important to nail these differences IMO.

So yeah, these differences are there. From our POV the unique capabilities of a given device (achievable through emulation) need to outweigh the complexity and effort of implementing the emulation of said device. Or at least I need to be super into some device, then I might not care about the effort 😆  (like with the upcoming AWE32 support).

I'm not even arguing, I think you and I are on the same page 😄 Thanks for stopping by.
